bokeh/io.py

```diff
@@ -393,14 +394,21 @@ def _detect_filename(ext):
     None if the script could not be found (e.g. interactive mode).
     """
     import inspect
-    from os.path import isfile, dirname, basename, splitext, join
+    from os.path import isfile, dirname, basename, splitext, join, curdir
 
     frame = inspect.currentframe()
     while frame.f_back and frame.f_globals.get('name') != '__main__':
         frame = frame.f_back
 
     filename = frame.f_globals.get('__file__')
-    if filename and isfile(filename):
+
+    if filename is None:
+        return tempfile.NamedTemporaryFile().name
+
+    elif not os.access(dirname(filename) or curdir, os.W_OK | os.X_OK):
+        return tempfile.NamedTemporaryFile().name
```

Comment:
I think this usage of `or` might be surprising to many python programmers. But in any case, why the fallback check to `curdir`? The branch below only ever uses `dirname(filename)` so the accessibility of `curdir` does not seem relevant?

As a personal preference, I don't use `elif` if all the branches `return`